File: graphs/0864-shortest-path-to-get-all-keys.py
from typing import List  # , Dict, Tuple
import unittest
from collections import deque

# https://leetcode.com/problems/shortest-path-to-get-all-keys/

# python3 -m unittest graphs/0864-shortest-path-to-get-all-keys.py


class Solution(unittest.TestCase):
    # # Approach #1: Brute Force DFS with memoization
    # # Time: O(m * n * 2^k * 4^(m*n)) - exponential due to exploring all paths
    # # Space: O(m * n * 2^k) for memoization
    # def shortestPathAllKeys(self, grid: List[str]) -> int:
    #     m, n = len(grid), len(grid[0])

    #     # Find start and count keys
    #     start_r, start_c = 0, 0
    #     key_count = 0

    #     for r in range(m):
    #         for c in range(n):
    #             if grid[r][c] == '@':
    #                 start_r, start_c = r, c
    #             elif 'a' <= grid[r][c] <= 'f':
    #                 key_count += 1

    #     all_keys = (1 << key_count) - 1
    #     inf = 1 << 31
    #     memo: Dict[Tuple[int, int, int], int] = {}

    #     def dfs(r: int, c: int, keys: int) -> int:
    #         if keys == all_keys:
    #             return 0

    #         if (r, c, keys) in memo:
    #             return memo[(r, c, keys)]

    #         result = inf

    #         for dr, dc in [(0, 1), (1, 0), (0, -1), (-1, 0)]:
    #             nr, nc = r + dr, c + dc

    #             if not (0 <= nr < m and 0 <= nc < n):
    #                 continue

    #             cell = grid[nr][nc]

    #             if cell == '#':
    #                 continue

    #             new_keys = keys

    #             if 'a' <= cell <= 'f':
    #                 new_keys = keys | (1 << (ord(cell) - ord('a')))

    #             if 'A' <= cell <= 'F':
    #                 if not (keys & (1 << (ord(cell) - ord('A')))):
    #                     continue

    #             sub_result = dfs(nr, nc, new_keys)
    #             if sub_result != inf:
    #                 result = min(result, 1 + sub_result)

    #         memo[(r, c, keys)] = result
    #         return result

    #     res = dfs(start_r, start_c, 0)
    #     return res if res != inf else -1

    # Approach #2: BFS with State Compression (Bitmask)
    # Time: O(m * n * 2^k) where m,n = grid dimensions, k = number of keys
    # Space: O(m * n * 2^k) for visited set
    def shortestPathAllKeys(self, grid: List[str]) -> int:
        m, n = len(grid), len(grid[0])

        # Find start position and count total keys
        start_r, start_c = 0, 0
        key_count = 0

        for r in range(m):
            for c in range(n):
                if grid[r][c] == '@':
                    start_r, start_c = r, c
                elif 'a' <= grid[r][c] <= 'f':
                    key_count += 1

        # BFS: (row, col, keys_mask, steps)
        queue = deque([(start_r, start_c, 0, 0)])
        visited = {(start_r, start_c, 0)}
        all_keys = (1 << key_count) - 1

        directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]

        while queue:
            r, c, keys, steps = queue.popleft()

            # Check if we have all keys
            if keys == all_keys:
                return steps

            # Explore 4 directions
            for dr, dc in directions:
                nr, nc = r + dr, c + dc

                # Check bounds
                if not (0 <= nr < m and 0 <= nc < n):
                    continue

                cell = grid[nr][nc]

                # Check if wall
                if cell == '#':
                    continue

                new_keys = keys

                # Check if it's a key
                if 'a' <= cell <= 'f':
                    new_keys = keys | (1 << (ord(cell) - ord('a')))

                # Check if it's a lock
                if 'A' <= cell <= 'F':
                    # Check if we have the key
                    if not (keys & (1 << (ord(cell) - ord('A')))):
                        continue

                # Check if state already visited
                if (nr, nc, new_keys) in visited:
                    continue

                visited.add((nr, nc, new_keys))
                queue.append((nr, nc, new_keys, steps + 1))

        return -1

    def test(self):
        for grid, expected in [
            (["@.a..", "###.#", "b.A.B"], 8),
            (["@..aA", "..B#.", "....b"], 6),
            (["@Aa"], -1),
            (["@...a", ".###A", "b.BCc"], 10),
        ]:
            output = self.shortestPathAllKeys(grid)
            self.assertEqual(expected, output, f"expected: {expected}, output: {output}")

File: graphs/0864-shortest-path-to-get-all-keys.go
package graphs

// https://leetcode.com/problems/shortest-path-to-get-all-keys/

// Approach: BFS with State Compression (Bitmask)
// Time: O(m * n * 2^k) where m,n = grid dimensions, k = number of keys
// Space: O(m * n * 2^k) for visited set
func shortestPathAllKeys(grid []string) int {
	m, n := len(grid), len(grid[0])

	// Find start position and count total keys
	startR, startC := 0, 0
	keyCount := 0

	for r := range m {
		for c := range n {
			if grid[r][c] == '@' {
				startR, startC = r, c
			} else if grid[r][c] >= 'a' && grid[r][c] <= 'f' {
				keyCount++
			}
		}
	}

	// BFS: [row, col, keys_mask, steps]
	type state struct {
		r, c, keys, steps int
	}

	queue := []state{{startR, startC, 0, 0}}
	visited := make(map[[3]int]bool)
	visited[[3]int{startR, startC, 0}] = true
	allKeys := (1 << keyCount) - 1

	directions := [][2]int{{0, 1}, {1, 0}, {0, -1}, {-1, 0}}

	for len(queue) > 0 {
		curr := queue[0]
		queue = queue[1:]

		// Check if we have all keys
		if curr.keys == allKeys {
			return curr.steps
		}

		// Explore 4 directions
		for _, dir := range directions {
			nr, nc := curr.r+dir[0], curr.c+dir[1]

			// Check bounds
			if nr < 0 || nr >= m || nc < 0 || nc >= n {
				continue
			}

			cell := grid[nr][nc]

			// Check if wall
			if cell == '#' {
				continue
			}

			newKeys := curr.keys

			// Check if it's a key
			if cell >= 'a' && cell <= 'f' {
				newKeys = curr.keys | (1 << (cell - 'a'))
			}

			// Check if it's a lock
			if cell >= 'A' && cell <= 'F' {
				// Check if we have the key
				if (curr.keys & (1 << (cell - 'A'))) == 0 {
					continue
				}
			}

			// Check if state already visited
			stateKey := [3]int{nr, nc, newKeys}
			if visited[stateKey] {
				continue
			}

			visited[stateKey] = true

			queue = append(queue, state{nr, nc, newKeys, curr.steps + 1})
		}
	}

	return -1
}
